Sources/SpeziBluetooth/CoreBluetooth/BluetoothPeripheral.swift

```diff
@@ -237,6 +237,34 @@ public actor BluetoothPeripheral: BluetoothActor { // swiftlint:disable:this typ
         }
     }
 
+    /// Establish a connection to the peripheral and wait until it is connected.
+    ///
+    /// Make a connection to the peripheral.
+    ///
+    /// - Note: This method waits until the connection has been established successfully. 
+    ///     Use ``connect()`` to skip waiting instead.
+    ///
+    /// - Note: You might want to verify via the ``AdvertisementData/isConnectable`` property that the device is connectable.
+    public func connectAndWait() async throws {
+        guard let manager else {
+            logger.warning("Tried to connect an orphaned bluetooth peripheral!")
+            return
+        }
+
+        await manager.connect(peripheral: self)
+
+        while true {
+            switch await waitForChange(of: \.state) {
+            case .connected:
+                return
+            case .connecting:
+                continue
+            case .disconnecting, .disconnected:
+                throw CBError(.connectionFailed)
+            }
+        }
```

Comment:
Instead of this loop-based approach I would propose the following.

I would just create a CheckedThrowing continuation and store the continuation object within the peripheral.
The continuation is resumed based on two events: `signalFullyDiscovered()` is called (resuming by returning void) or a connection error is received (resuming by throwing an error). In the erroneous case, the `didFailToConnect` delegate method of the `BluetoothManager` is getting called (see [here](https://github.com/StanfordSpezi/SpeziBluetooth/blob/8e94fc71720ef3fcf7f5d9dba9eef603c5151d7a/Sources/SpeziBluetooth/CoreBluetooth/BluetoothManager.swift#L704)).

You can have a look at, e.g., the `readRSSI()` how the continuation pattern is roughly implemented. Of course this is not quite the same.

The `BluetoothPeripheral` already has a `handleDisconnect()` that handles state cleanup and is called both for a failed connection and a regular disconnect. `handleDisconnect()` is getting called from `BluetoothManager/discardDevice(device:)` which again is called from the `didFailToConnect` and the `didDisconnectPeripheral` delegate methods. I would extend both methods to accept an optional error object and adjust the `handleDisconnect()` implementation to resume the "connectContinuation" with the error object. I think that should be the easiest way to get access to the CoreBluetooth connection error.

## Special Cases

There are two special cases I can currently think of one would need to address:

* **Second `connect()` call**: What happens if a connect() method is called while the peripheral is already connecting (e.g., another continuation was created). Throw an error? Or just "queue" and return the same result as for the first continuation?
* **Task cancellation**: Task cancellation is cooperative in Swift. We would need to wrap the continuation inside a [withTaskCancellationHandler(operation:onCancel:)](https://developer.apple.com/documentation/swift/withtaskcancellationhandler(operation:oncancel:)) and cancel the connection by just calling the disconnect() method of the peripheral. That should handle everything automatically.

Reply:
I tried this approach - thank you for the guidance! Please have another look, whether this is what you intended 😊 I'm especially not sure about the implementation inside the `connect()` method itself (incl the `assumeIsolated` and the disconnect task).